github/workflows: Build test images

luigix25 opened the pull request:
Build test images in the CI to make sure that a PR doesn't break tests.

Suggested-by: Stefano Garzarella <sgarzare@redhat.com>

stefano-garzarella commented on .github/workflows/rust.yml:
```diff
@@ -20,7 +20,9 @@ jobs:
       - name: Install rust toolchain
-        run: rustup toolchain install --profile minimal
+        run: |
+          rustup +nightly target add x86_64-unknown-none
```
Can we avoid this and use the `nightly-check` pipeline to run `make test-igvm` since we already install the nightly toolchain in that pipeline?

luigix25 replied:
> Can we avoid this and use the `nightly-check` pipeline to run `make test-igvm` since we already install the nightly toolchain in that pipeline?

In that case I'd need to install the TPM build dependencies in that pipeline. Do you still prefer to move it to the other pipeline?

luigix25 commented on kernel/src/cpu/vc.rs:
```diff
@@ -338,11 +338,13 @@ mod tests {
-        // SAFETY: CPUID does never affect safety.
-        let extended_info = unsafe { __cpuid_count(CPUID_EXTENDED_FUNCTION_INFO, 0) };
-        assert!(extended_info.eax >= CPUID_MEMORY_ENCRYPTION_INFO);
+        if !is_test_platform_type(SvsmPlatformType::Native) {
```
As you prefer. I tried to stick to what others did in the code

stefano-garzarella commented on Makefile:
```diff
@@ -4,7 +4,15 @@ SVSM_ARGS += --features ${FEATURES}
 XBUILD_ARGS += -f ${FEATURES}
 endif
 
+#We don't support virtio in nocc mode
+ifeq ($(NOCC),1)
+NOCC=--nocc
+FEATURES_TEST ?= vtpm
+else
+NOCC=
 FEATURES_TEST ?= vtpm,virtio-drivers
+endif
+
```
That we don't use in the CI, so I'd avoid adding new parameters when you have a way to do that.
Also, IIUC the missing part is just #775 that I hope we will merge soon.

luigix25 replied:
yep, that is the only missing part. I'll remove this part.

stefano-garzarella commented:
> update: if some of the tests doesn't pass (and panics the kernel) the GH action will run until timeout (180s). Investigating..

@luigix25 IIRC @osteffenrh added the `timeout` in `scripts/test-qemu-nocc-boot.sh` because QEMU was never exiting with just the boot. But with `test-in-svsm` IIRC, QEMU will exit at the end of the tests, so maybe we don't need that script anymore. (And if we need, maybe we need to update the name)

luigix25 commented:
> > update: if some of the tests doesn't pass (and panics the kernel) the GH action will run until timeout (180s). Investigating..
> 
> @luigix25 IIRC @osteffenrh added the `timeout` in `scripts/test-qemu-nocc-boot.sh` because QEMU was never exiting with just the boot. But with `test-in-svsm` IIRC, QEMU will exit at the end of the tests, so maybe we don't need that script anymore. (And if we need, maybe we need to update the name)

If everything goes well, then `test-in-svsm` exits correctly, but if something goes wrong it panics and qemu remains open. I'm modifying Oliver's script to detect a panic.

stefano-garzarella commented:
> > > > update: if some of the tests doesn't pass (and panics the kernel) the GH action will run until timeout (180s). Investigating..
> > > 
> > > 
> > > @luigix25 IIRC @osteffenrh added the `timeout` in `scripts/test-qemu-nocc-boot.sh` because QEMU was never exiting with just the boot. But with `test-in-svsm` IIRC, QEMU will exit at the end of the tests, so maybe we don't need that script anymore. (And if we need, maybe we need to update the name)
> > 
> > 
> > If everything goes well, then `test-in-svsm` exits correctly, but if something goes wrong it panics and qemu remains open. I'm modifying Oliver's script to detect a panic.
> 
> Strange, in the panic handler we have `platform::halt();` (
> 
> https://github.com/coconut-svsm/svsm/blob/579de9f3ead1d7fc596965d818f465979e00bea1/kernel/src/svsm.rs#L422
> ), so I'm expecting QEMU to exit on `hlt`, no?

no, so maybe we should call `testing:exit()`, we need to write in the `QEMU_EXIT_PORT`

luigix25 commented:
> > 
> > Strange, in the panic handler we have `platform::halt();` (
> > https://github.com/coconut-svsm/svsm/blob/579de9f3ead1d7fc596965d818f465979e00bea1/kernel/src/svsm.rs#L422
> > 
> > ), so I'm expecting QEMU to exit on `hlt`, no?
> 
> no, so maybe we should call `testing:exit()`, we need to write in the `QEMU_EXIT_PORT`

Exactly, you have to write in the `QEMU_EXIT_PORT`. You mean calling it every time the kernel panics?

osteffenrh commented on kernel/src/testing.rs:
```diff
@@ -92,9 +94,19 @@ pub fn svsm_test_runner(test_cases: &[&test::TestDescAndFn]) {
+                    .unwrap();
+            }
+            SvsmPlatformType::Native => {
+                SVSM_PLATFORM.get_io_port().outl(QEMU_EXIT_PORT, 0);
```
I know this value was not chosen in this PR, but I think we should use something else here, since this causes QEMU to exit with code `1`, which is also used for general (startup) errors. AFAIK `exit_code = (port_value << 1) | 1`. See https://os.phil-opp.com/testing/#exiting-qemu  (Not sure where the actual QEMU documentation for this is).
But we can address this is a separate PR (if we want to change it).

luigix25 replied:
I like this idea too! Thanks for the suggestion!

stefano-garzarella commented:
@luigix25 we just merged #775 so I think we can now rebase and enable also `virtio-drivers` feature, right?

luigix25 commented:
> @luigix25 we just merged #775 so I think we can now rebase and enable also `virtio-drivers` feature, right?

Thanks! yep, I am already working on this and some minor fixes

stefano-garzarella commented on kernel/src/utils/util.rs:
```diff
@@ -86,6 +89,22 @@ pub unsafe fn zero_mem_region(start: VirtAddr, end: VirtAddr) {
+pub const QEMU_EXIT_FAIL: u32 = 0x11;
+
+#[cfg(test)]
+pub fn qemu_write_exit(value: u32) {
```
What about using an `enum` as parameter, and here do the mapping to u32 values that QEMU expects?

stefano-garzarella commented on Makefile:
```diff
@@ -31,6 +31,12 @@ else ifeq ($(V), 2)
 CARGO_ARGS += -vv
 endif
 
+ifeq ($(NOCC),1)
+NOCC=--nocc
+else
+NOCC=
+endif
+
```
Can we just add a parameter that we can eventually reuse in the future, instead of adding custom parameter for every new one.

I mean something like `CLIPPY_ARGS` or `FEATURES_TEST`, in this case we can name it something like `TEST_ARGS`.

stefano-garzarella commented on scripts/test-in-svsm.sh:
```diff
@@ -47,9 +47,37 @@ dd if=/dev/urandom of="$TEST_DIR/svsm_state.raw" bs=512 count=1024
+done
+
+
+if $SCRIPT_DIR/launch_guest.sh --igvm $SCRIPT_DIR/../bin/coconut-test-qemu.igvm \
```
Also, what is the difference between this branch and the next one (33 exit value)?

luigix25 replied:
> Can we execute the command, then check `$?` ?

Not really. QEMU returns a non-zero value for success and because of the `-e` option it would stop executing.
`|| true` would overwrite the return value with 0.

So that if is a trick to execute the script and to check the result, The "side-effect" is an unused branch.
I really didn't want to disable `-e` and enable it again afterwards.

stefano-garzarella replied:
What about something like this?

```diff
diff --git a/scripts/test-in-svsm.sh b/scripts/test-in-svsm.sh
index b63aec0d..17b597d3 100755
--- a/scripts/test-in-svsm.sh
+++ b/scripts/test-in-svsm.sh
@@ -63,13 +63,13 @@ while [[ $# -gt 0 ]]; do
 done
 
 
-if $SCRIPT_DIR/launch_guest.sh --igvm $SCRIPT_DIR/../bin/coconut-test-qemu.igvm \
+$SCRIPT_DIR/launch_guest.sh --igvm $SCRIPT_DIR/../bin/coconut-test-qemu.igvm \
     --state "$TEST_DIR/svsm_state.raw" \
     --unit-tests $TEST_DIR/pipe \
-    $NOCC ; then
-    exit_value=0
+    $NOCC || svsm_exit_code=$?
+
 # SVSM returns 33 when everything goes well
-elif [ $? -eq 33 ]; then
+if [ $svsm_exit_code -eq 33 ]; then
     echo "All tests passed"
     exit_value=0
 else
```

luigix25 replied:
it works! Thanks for the hint!

stefano-garzarella commented on .github/workflows/qemu.yml:
```diff
@@ -106,20 +106,18 @@ jobs:
-      - name: Build SVSM
-        run: cargo xbuild configs/qemu-target.json
+      - name: Build test
+        run: NOCC=1 make bin/coconut-test-qemu.igvm
```
Why we need `NOCC=1` here?

luigix25 replied:
oops, leftover

stefano-garzarella commented on kernel/src/testing.rs:
```diff
@@ -90,12 +88,7 @@ pub fn svsm_test_runner(test_cases: &[&test::TestDescAndFn]) {
-            .ioio_out(QEMU_EXIT_PORT, GHCBIOSize::Size32, 0)
-            .unwrap();
-    }
+    qemu_write_exit(QEMUExitValue::Success);
```
Okay, re-reading @osteffenrh motivation in https://github.com/coconut-svsm/svsm/pull/823#discussion_r2394017961 made it much more clear about the reason you're changing it.

LGTM (using something different from 0), but please add that motivation in the commit message.

stefano-garzarella commented on kernel/src/utils/util.rs:
```diff
@@ -6,6 +6,9 @@
 
 use crate::address::{Address, VirtAddr};
 use crate::types::PAGE_SIZE;
+#[cfg(test)]
```
So maybe you can use ` kernel/src/testing.rs`  and do:
```
    loop {
        debug_break();
        #[cfg(test)]
        qemu_write_exit(QEMUExitValue::Fail);
        platform::halt();
    }
```

Like we do in svsm_main() when calling test_main().

luigix25 commented:
Addressed Stefano's comments:

- I moved `qemu_write_exit` to `kernel/src/testing.rs`
- Removed the stub, it is not necessary anymore.
- Refactored `qemu_write_exit`: there is no need to differentiate per platform anymore.
- Added comment explaining the QEMU value
- Introduced a commit for QEMU exit value

stefano-garzarella commented on kernel/src/testing.rs:
```diff
@@ -99,8 +99,8 @@ fn exit() -> ! {
 pub enum QEMUExitValue {
-    Success = 0x0,
-    Fail = 0x1,
+    Success = 0x10,
```
Please, move this change before the refactoring. I'd avoid adding new code (e.g.  `enum QEMUExitValue`) in a commit and change it in a following commit of the same PR.

stefano-garzarella commented on kernel/src/testing.rs:
```diff
@@ -108,8 +106,8 @@ pub enum QEMUExitValue {
-        current_ghcb()
-            .ioio_out(QEMU_EXIT_PORT, GHCBIOSize::Size32, value as u64)
-            .unwrap();
+        SVSM_PLATFORM
```
Ditto, move this change before the refactoring.

stefano-garzarella commented on kernel/src/svsm.rs:
Please fix https://github.com/coconut-svsm/svsm/pull/823/commits/f67c1e96236298f44688b63557f75878cfb3bfe6 description. From the title, it seems it does only the refactoring, while the real change here is the qemu_write_exit call the panic handler. I think the title should talk more about that than the refactoring. (Or split this in 2 commits, but I'm fine also with a single commit with a title fixed).

stefano-garzarella commented on scripts/test-in-svsm.sh:
```diff
@@ -47,9 +47,38 @@ dd if=/dev/urandom of="$TEST_DIR/svsm_state.raw" bs=512 count=1024
+
+# SVSM writes 0x10 to the QEMU exit port when all tests passed.
+# This results in QEMU returning 0x21 ((0x10 << 1) | 1)
+if [ $svsm_exit_code -eq 33 ]; then
```
What about using the hex value also in the code?
e.g. `if [[ $svsm_exit_code -eq 0x21 ]]; then`

stefano-garzarella commented on .github/workflows/qemu.yml:
```diff
@@ -106,20 +106,18 @@ jobs:
 
       ########################
 
       - name: Run SVSM in QEMU
```
```suggestion
      - name: Run test-in-svsm in QEMU
```

luigix25 commented:
- Reordered commits according to Stefano's suggestions.
- Reworded a commit message
- Very minor changes to bash and yaml

stefano-garzarella commented on kernel/src/testing.rs:
```diff
@@ -87,10 +87,19 @@ pub fn svsm_test_runner(test_cases: &[&test::TestDescAndFn]) {
 
+#[repr(u32)]
+#[derive(Debug)]
+pub enum QEMUExitValue {
```
Please move this to the next commit, here just fix the issue.
In general, we avoid mixing new code with fixes, to avoid problems with backports (it's true, we don't have stable branches yet, but it's good practice to follow).

stefano-garzarella reviewed (commented):
LGTM, just a minor comment.

stefano-garzarella reviewed (approved):
@joergroedel I think we can merge this, but I'll wait for your review.

Note the CI time is not affected by this PR, since running tests is just 2 sec (see `Run test-in-svsm in QEMU` in https://github.com/coconut-svsm/svsm/actions/runs/18306479849/job/52124795547?pr=823)
